linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1726,6 +1770,102 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
+
+    if (!targetMethod.attributes.inlineable) {
+      treeNotInlined
+    } else {
+      def getImport(target: ImportTarget): Option[JSNativeLoadSpec.Import] = target match {
+        case ImportTarget.Class(className) =>
+          getJSNativeImportOf(className)
+        case ImportTarget.Member(className, methodName) =>
+          getJSNativeImportOf(className, methodName)
+      }
+
+      val maybeImportTarget = targetMethod.attributes.jsDynImportInlineTarget.orElse {
+        targetMethod.attributes.jsDynImportThunkFor.flatMap { thunkTarget =>
+          val id = staticCall(className, MemberNamespace.Public, thunkTarget)
+          if (id.attributes.inlineable)
+            id.attributes.jsDynImportInlineTarget
+          else
+            None
+        }
+      }
+
+      val maybeInlined = for {
+        importTarget <- maybeImportTarget
+        jsNativeLoadSpec <- getImport(importTarget)
+      } yield {
+        pretransformExprs(args) { targs =>
+          tryOrRollback { cancelFun =>
+            val moduleName = freshLocalNameWithoutOriginalName(
+                LocalName("module"), mutable = false)
+
+            val moduleParam =
+              ParamDef(LocalIdent(moduleName), NoOriginalName, AnyType, mutable = false)
+
+            val importReplacement = ImportReplacement(importTarget, moduleName, jsNativeLoadSpec.path,
+                newSimpleState(Unused), cancelFun)
+
+            val (newCaptureParamsWithUse, newBody) = {
+              val MethodDef(_, _, _, params, resultType, Some(body)) = getMethodBody(targetMethod)
+              transformIsolatedBody(Some(targetMethod), NoType, params, resultType, body,
+                  scope.implsBeingInlined, Some(importReplacement))
+            }
+
+            if (!importReplacement.used.value.isUsed)
+              cancelFun()
+
+            withCaptures(newCaptureParamsWithUse, targs) { (captureParams, captureValues, cont1) =>
+              val inlinedClosure = Closure(arrow = true, captureParams, List(moduleParam),
+                  restParam = None, newBody, captureValues)
+
+              def genThen(receiver: Tree, expr: Tree) =
+                JSMethodApply(receiver, StringLiteral("then"), List(expr))
+
+              val importTree = config.coreSpec.moduleKind match {
+                case ModuleKind.NoModule =>
+                  throw new AssertionError("ApplyDynamicImport in NoModule mode")
+
+                case ModuleKind.ESModule =>
+                  JSImportCall(StringLiteral(jsNativeLoadSpec.module))
+
+                case ModuleKind.CommonJSModule =>
+                  val require =
+                    JSFunctionApply(JSGlobalRef("require"), List(StringLiteral(jsNativeLoadSpec.module)))
+
+                  val unitPromise = JSMethodApply(
+                    JSGlobalRef("Promise"), StringLiteral("resolve"), List(Undefined()))
+                  genThen(unitPromise, Closure(arrow = true, Nil, Nil, None, require, Nil))
+              }
```

Comment:
I'd like to factor this out, but couldn't really wrap my head around how tree generation helpers are called in OptimizerCore. Seems like objects with apply methods are most common (e.g. `AndThen`).

Should I define an `object JSImport` with an apply method that does this? (and `object Then`?)

Reply:
That sounds reasonable. I guess we went the `apply` route because we sometimes need the `unapply` as well. Perhaps we even had `AndThen.unapply` in the past, and it became irrelevant later?

In any case, defining a new object near `AndThen` shouldn't hurt.